Hooks: Add pre-commit hook for PHPCS.

kwight opened the pull request:
Now that we have PHP files in our `/apps` directory, let's have PHPCS run automatically as a pre-commit hook.

It requires that PHPCS is available locally, and will not attempt to process PHP files without it (but will continue with linting everything else). The custom ruleset defined at `apps/phpcs.xml` is used for both PHPCBF and PHPCS.

#### Testing instructions

* Apply this branch locally.
* Modify a PHP file within `/apps/full-site-editing` that includes linting problems (eg. too many spaces, unused variables).
* From the command-line at your `wp-calypso` root, run `node bin/pre-commit-hook.js`.
* Verify that the affected file was cleaned up, and issues were reported.
* Do funny things; try to break it, or cause unexpected behaviour.

kwight committed:
Hooks: Add pre-commit hook for PHPCS.

gwwar commented:
Let us know if you need help with anything here @kwight! Is this ready for testing, or are there particular problems left to work through? 

This is 💯to add linting checks vs needing to remind folks in reviews / constantly tidying! Thanks for taking a look.

kwight committed:
Ensure valid defaults for empty `groupBy` results.

kwight committed:
Temporarily allow process to continue when `phpcbf` returns a non-zero exit code
(like when it successfully fixes something).

kwight committed:
PHPCBF returns a `0` exit code when it had nothing to fix, a `1` exit
code when it successfully fixes issues, and a `2` exit code on actual
failures of some sort.

kwight committed:
Only process PHP files if the proper executables are locally installed.

kwight committed:
Ensure PHPCBF files are staged.

kwight committed:
Run PHPCS.

kwight committed:
Add comment explaining PHPCS exit codes.

kwight committed:
Stick to the same `spawnSync` as the others, for easier `stdout`
handling.

kwight committed:
Have the PHPCS tools use our custom ruleset.

vindl commented:
I'm running into the following issue during testing:

`PHP files will not be processed because PHPCS was not found.`

although I have `phpcs` installed and it's working for FSE folder when I run it manually. Might be something with my local setup. 🤔Tried replacing `which` with `hash`, `type`, and `command` but that didn't help.

I also found this output to be odd/unexpected:

```
Prettier formatting staged file: apps/full-site-editing/full-site-editing-plugin/full-site-editing/class-full-site-editing.php
[error] No parser could be inferred for file: apps/full-site-editing/full-site-editing-plugin/full-site-editing/class-full-site-editing.php
child_process.js:656
    throw err;
    ^
```

vindl commented:
The above errors were caused by not having `phpcs` in my path (hat tip @kwight).

vindl commented on bin/pre-commit-hook.js:
```diff
@@ -38,6 +38,19 @@ function linterFailure() {
+		execSync( 'which -s phpcs phpcbf' );
+		return true;
+	} catch ( error ) {
+		console.log( chalk.red( 'PHP files will not be processed because PHPCS was not found.' ) );
```
For follow up: a link to docs with setup instructions could be helpful here.

kwight replied:
https://github.com/Automattic/wp-calypso/issues/34552

vindl reviewed (approved):
LGTM! :shipit:
